samples/common/common.js

```diff
@@ -0,0 +1,118 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * Common functions/classes for separate sample handling.
+ */
+
+'use strict';
+
+// A builder for individual sample files. This allows a sample to provide
+// some simple metadata that may be passed to yargs in various contexts,
+// specifically when running with the sample as main, versus using one
+// of the sample runner files.
+class SampleDeclaration {
+  constructor(yargs) {
+    this.name = null;
+    this.helpText = '';
+    this.argString = '';
+    this.spacedArgs = '';
+    this.extraOptions = {};
+    this.exampleText = [];
+    this.yargs = yargs;
+  }
+
+  // Pass the name of the command (e.g. "create-push").
+  commandName(name) {
+    this.name = name;
+    return this;
+  }
+
+  // Pass the help text for the command (e.g. "Does the foo.").
+  help(text) {
+    this.helpText = text;
+    return this;
+  }
+
+  // Pass the argument string and any extra options. These are passed
+  // directly to yargs.
+  args(argString, extraOptions) {
+    this.argString = argString;
+    this.spacedArgs = this.argString ? ` ${this.argString}` : '';
+    this.extraOptions = extraOptions || {};
+    return this;
+  }
+
+  // Pass an example string for use as a yargs example.
+  example(exampleText) {
+    this.exampleText = this.exampleText || [];
+    this.exampleText.push(exampleText);
+    return this;
+  }
+
+  // Call this last; it will either run the sample with what's been
+  // given (if it's running as main) or export this object for the
+  // consolidated runner module to use. 'forceMain' is for testing
+  // purposes.
+  execute(sampleModule, executionCallback, forceMain = false) {
```

Comment:
Is there a new sample requirement to allow the samples to be runnable via cli and requirable? I like it, I just haven't seen anything similar in other modules. If that is the case it might be worth bundling this up as its own module that we can import in other projects.

Reply:
This is another point where I made some creative interpretation. The style guide says the samples should be in separate files so they can be run independently, but I felt like it was important to keep from pulling the rug out from under users, so I wanted to make the original topics.js and subscriptions.js keep working.

The other consideration there is just that some of the samples require numeric and/or optional arguments, and it was a lot easier to keep using yargs for that. And if I'm doing that anyway, I might as well make it plug into those original files.

Reply:
I see. I am definitely all for backwards compatibility, but I haven't observed that happening in other repos that have split the samples. I would be curious to hear @JustinBeckwith's take on this

Reply:
Me too! It's pretty easy to just take out the aggregate sample files if we want to simplify it, but I might still be interested in keeping the common argument handling even if we do that (because of the numeric args and help text).

Reply:
I re-read the style guide again and it actually says we're deprecating aggregate sample runners that have full command lines. Basically the emphasis is on making the samples easily paste-capable in the docs site. So this might be overengineered at this point. I'll bug Justin about it.

Reply:
We had a talk internally and decided to go ahead and remove the aggregate sample runners, and fully simplify this all down. I've just changed the system-test modules to run the individual samples directly.